### DISCUSSION TITLE
Not able to emit from background thread. Similar to: https://github.com/miguelgrinberg/Flask-SocketIO/issues/434

### DISCUSSION THREAD

**Discussion by Aditya-Dexian** [miguelgrinberg/Flask-SocketIO#2073] [Q&A] [answered]
Hi,

This problem has been described in https://github.com/miguelgrinberg/Flask-SocketIO/issues/434 as well.

I am a beginner programmer trying to implement a real-time transcription solution using Python Flask, React.js, and websockets.

I am asking for help with my code. 

I seriously appreciate any and all help with this query.

_Note: This code is minimal, not my entire code_

server.py:
```
# this one sends the transcript properly and properly sets it. this is how i know the frontend can definitely pick up these requests properly.
```
   
I would really appreciate all the help I can get. Thank you guys so much in advance.

**Comment by Aditya-Dexian** [comment]
Miguel responded to me in another place, suggesting that I enable logging for debugging purposes. I am looking into the docs currently to enable that. Thank you.

**Reply by Aditya-Dexian** [reply]
![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/170400647/96833ab0-1d1d-4bb0-86ca-d059c6f422ca)
Ok this is a screenshot of what logging I see in my terminal.
The top stuff is some handshake stuff I think, there is one where I collect emails from the frontend which goes through. There is one where I press the start button on front end, it goes through.
The boxes with the result, emitted transcript, transcription is me trying to send a transcript_update request from backend through azure transcription stuff.
The other boxes with the success code 200 is when I refresh my localhost:5000/send.

**Reply by miguelgrinberg** [reply]
You are showing me your own logs, or the logs from the web server. I need to see the Socket.IO logs. See the troubleshooting section of the docs to learn how to enable Socket.IO logs.

**Reply by Aditya-Dexian** [reply]
Hi Miguel, 
thanks for your response.
I am trying to use [these docs](https://flask-socketio.readthedocs.io/en/latest/getting_started.html#debugging-and-troubleshooting) to learn how to debug. At the bottom it says I can do:
`socketio = SocketIO(logger=True, engineio_logger=True)`
to put logs to sterr but I am having trouble using this line. I get these errors:

```
Traceback (most recent call last):
  File "C:\backend\server.py", line 109, in <module>
    main()
  File "C:\backend\server.py", line 104, in main
    socketio.run(app)
  File "C:\.call-center\Lib\site-packages\flask_socketio\__init__.py", line 631, in run
    if self.server.eio.async_mode == 'threading':
       ^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'eio'
```

My code for setting up the socket:
```
app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
CORS(app,resources={r"/*":{"origins":"*"}})    # my react frontend is on port 3000 so im using this cors stuff to enable that origin
# socketio = SocketIO(app, cors_allowed_origins="*")
socketio = SocketIO(logger=True, engineio_logger=True)
# socketio = SocketIO(app, async_mode='threading', cors_allowed_origins="*")

if __name__ == '__main__':
     socketio.run(app)

```

I'm on version 5.3.6 of Flask-SocketIO.

I must be misunderstanding the docs. 

Thank you for your continued time and consideration.

**Reply by miguelgrinberg** [reply]
The logging arguments have to be added to the `SocketIO` class. They are not supposed to replace your own arguments. Try this:

    socketio = SocketIO(app, async_mode='threading', cors_allowed_origins="*", logger=True, engineio_logger=True)

**Reply by Aditya-Dexian** [reply]
Yes, this works.
![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/170400647/5096ca85-190e-42f1-b2c1-85ead8f96f27)
I clicked start recording button sending emit from frontend, received in flask.
did the send() request from localhost:5000/send
talked into my mic for transcription, shows that transcript_update is being emitted to default namespace /. this was not picked up by the front end which is still stuck showing "hello" as the last transcript update.
Again thank you for your continued time and consideration

**Reply by miguelgrinberg** [reply]
Is this application a single page app? Or in other words, how does the browser request the `/send` page? If you are doing normal navigation your WebSocket connection is going to drop. The browser disconnects when the page changes, with Flask-SocketIO you need to write your entire application in a single page for the connection to stay alive.

**Reply by Aditya-Dexian** [reply]
```
@app.route('/send')
def send():
    socketio.emit('transcript_update', {"transcript": "hello"})
    return "hello"
```
This code segment is in the same file that I am using to host the Flask app. I am manually going to http://localhost:5000/send to trigger this request. 

I attached an image where I try to emit the transcript through Azure first in case somehow the /send request was somehow messing it up.

![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/170400647/c5eb5008-b194-411e-98e1-5f1b643e694d)

**Reply by Aditya-Dexian** [reply]
My React frontend uses only one page:
I render App.js:
```
import AgentView from './AgentView';
import Menu from './Menu'

export default function App() {
  const [menuActive, setMenuActive] = useState(true)
  const [agentViewActive, setAgentViewActive] = useState(false)
  function disableMenu() {    
    setMenuActive(false);
    setAgentViewActive(true);
  }

  return (
    <div id='page'>
      <Menu active={ menuActive } disableMenu={disableMenu}/>
      <AgentView active={ agentViewActive }/>
    </div>
  );
};
```

I use Menu.js to collect emails:
```
import React, { useEffect, useState } from "react";
import './Menu.css';
import socketIOClient from 'socket.io-client';

let socket = socketIOClient("http://localhost:5000/");

export default function Menu({ active, disableMenu }) {
    useEffect(() => {
        if (!active) {
            socket.close();
            console.log('menu socket closed');
        }
    }, [active])

    function onMenuClick(email1, email2) {
        socket.emit('update_emails', { "email1": email1, "email2": email2});
        disableMenu()
    }

    return (
        <div id='menu-container' className={active ? '' : 'hidden'}>            
            <div id='menu-body'>
                    <input for email1>
                    <input for email2>
                    <button onClick={ () => onMenuClick(email1, email2) }>Submit</button>
            </div>
        </div>
    )
}
```
I attached AgentView.js earlier (referred to as react.js). 

I was struggling with creating a socket instance in App.js and then passing it down to the Menu and AgentView elements, so I just chose to create one in each file that needed it (two). I think this is not a very good solution, but I think it should still be working. Even after I close the menu socket after rendering menu inactive, the agentview socket is still able to pick up requests (as shown by the start_recording and the /send hello transcript requests). 

Additionally, if I just comment out the <Menu> line in App.js to make it so that the Menu socket instance is never created, the same issue still persists where the transcript request from the azure transcriber stuff does not work but the transcript(hello) request from the /send route does.

Thank you for your continued time and assistance.

**Reply by Aditya-Dexian** [reply]
Further test..
Maybe I misunderstood your last comment about the application being on single page. I was thinking you were referring to the frontend but maybe you were referring to the backend?

server.py (flask host file):
```
opening/refreshing http://localhost:5000 will do /send request. 
trying to test stuff without all of the threads.

/send -> transcribers.test() -> return "result" -> emitting this returned value -> properly received by the front end

uncommenting the # server.emit_transcript("result") -> emit_transcript function is called with "result" parameter -> transcript_update request is sent -> not received

**Reply by miguelgrinberg** [reply]
I'm sorry, but I think you do  not understand what I'm saying. You shared this log:

![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/170400647/5096ca85-190e-42f1-b2c1-85ead8f96f27)

The top of the log shows that you had a Socket.IO connection. Then the client navigated to `/send`. And after that there are no more Socket.IO logs, which means that the connection disappeared. This is because the browser drops all connections when it goes to another page, so you should design your application so that the page does not change.

All the activity that appears in your log after `/send` was accessed are the emits that you say don't work. But they are working just fine. The problem is that you are emitting from a server that has no active connections to any client.

**Reply by Aditya-Dexian** [reply]
Thanks for your response.

![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/170400647/ea944afe-9e9f-47e5-964e-6e5e38dfb097)

I removed the /send thing completely from my code so I am pretty sure I am now not going to any other pages. 

I think you are saying that the websocket connection is somehow being dropped. To my understanding, if it is dropped, I should not be able to make and receive any more requests after the fact.

Referring to the new screenshot:
1: upgrade to socket successful, pass email input from front end to back end (successful)
2. press start button, call start recording function in backend from the frontend (successful)
3. I say hello, transcriber picks it up, tries to shoot transcript_update, sentiment_update from backend to frontend. i don't see any updates on the front end like I should and like I did earlier when using the /send (unsuccessful)
4. press stop button, call stop recording function in backend from frontend. python errors because I have not implemented this function properly but the function is still being called (successful)

Please correct me if I'm wrong but even after the /send stuff is removed, I am still able to make a successful websocket request which implies the socket connection is not broken?

Once again, thank you for your continued time and assistance.

**Reply by miguelgrinberg** [reply, marked as answer]
The lines that start with 127.0.0.1 comes from the web server. They show connections that ended. The ones that have `transport=websocket` in the URL are Socket.IO connections that ended. I really can't tell for sure because you are using many connections in your client, but at least some connections are going away for some reason.

So first you need to figure out why you have many connections instead of just one. And second why they are getting closed.

**Reply by Aditya-Dexian** [reply]
Hi Miguel,

Thank you for your response and help.

I pondered upon your last reply for a while, and I thought maybe the way I've written my code to import each other is the issue. When I run server.py, it imports transcribers.py which imports server.py again, causing some weird issues. 

I confirmed my issue by recombining all my files into one large 500+ line file and transcription ran as desired.

Now I realized exactly my issue, and I wrote the following code:

_This code is separate from the stuff I've shared above._

 _server.py:_
```

```

_SocketConfig.py:_
```
```

_transcribers.py:_
```
import server

note how importing server nolonger runs any of the stuff like app=Flask(__name__), socketio = SocketIO(app, ........)
```

This works.

Looking at the logs, I only see one socket connection (I think) sending the requests.

I seriously couldn't have gotten this far without you.

Thank you once again Miguel.
